### ISSUE TITLE
Consider using postgres ltree module for metric paths

### ISSUE THREAD

**Issue by starstuck** [dalmatinerdb/dqe_idx_pg#32]
I would like to suggest some research into benefits of switching metric path from array to 'label tree' data structure provided by optional postgresql module. That module provides extensive facilities for searching through label trees which could be especially useful for pulling one level of metrics (metric tree traversal).

If that lands in indexer we could later on easily build full blown path filtering based on querying facilities provided by ltree module.

**Comment by starstuck** [comment]
Here is link to documentation showing query flexibility of that data structure: https://www.postgresql.org/docs/9.6/static/ltree.html

**Comment by Licenser** [comment]
looks interesting, I got to read up to say more :) but we always can do dqe_idx_pg_ltree as a own resolver to experiment :)

**Comment by Licenser** [comment]
reading over it I don't think it'll work, the problem comes that labels have to fit a subset of characters, and are dot seperted. While we allow any kind of character in the path and the dot for seperation is only a readability thing (we could use any other character if desired

**Comment by starstuck** [comment]
Probably those are limitations we could overcome. We could escape, or encode paths (using url encoding for example) to make sure only valid characters are stored in db.

I will try to set up some experiment to see whether performance coming from that index is worth the effort. I'll keep you posted when I get to some numbers we can use for comparison.

**Comment by Licenser** [comment]
hah I totally dind't think of that, base64 or something would indeed solve the issue, not human readable but that is a bit of a secondary thing :)!

**Comment by starstuck** [comment]
After spending more time with it, I think the problem is not with data structures, but with properties of particular index.

Metrics are using btree index, which does not handle matching by parts, but it supports ordering and comparison. I used technique used by you in getting quick list of collections, the `WITH RECURSVE ...` queries. That allowed to efficiently force postgres to start wit top most node and then just skip through nodes on interesting levels. If you are interested in details, they are in commit https://github.com/dalmatinerdb/dqe_idx_pg/commit/8a19b4dd4c9114fd30d05b66188e988e77f0ef3f

I think that is really good for what we need now. It is on a branch where I made some further changes to make efficient use of postgres hstore data type for keeping metric dimensions. I still need to do few more checks to confirm it works well. I will make pull request for that shortly.

If it came down to having more flexible queries based on part of path, it can still be used with arrays and GIN index. That index will support for metrics having certain elements in path.

I am closing this issue as dead end.
